dts: zynqmp: Add proper DT-based integration of On-Chip Memory

ibirnbaum opened the pull request:
Just as the Xilinx Zynq-7000, the ZynqMP also comes with a 256 kB On-Chip Memory area.

Until now, this memory area is already properly configured for use as DMA memory when the MPU is enabled, but any use of this memory requires the use of raw pointers.

Identical to how the OCM has been integrated for the Zynq-7000, add a declaration for the ZynqMP's OCM to the device tree, so that the existing linker command file will place any data to whose declaration "__ocm_bss_section" or "__ocm_data_section" has been added there. For the OCM's DT node, the Zynq-7000's compatibility identifier "xlnx,zynq-ocm" is re-used, as it also matches the ZynqMP.

Supporting modifications:
* Add a matching 'chosen' entry to the device tree of the target qemu_cortex_r5 which is based on the ZynqMP.
* Remove the distinction between Zynq-7000 and ZynqMP for the GEM Ethernet driver, which until now has already been using the Zynq-7000's OCM as described above, but has been using regular memory on the ZynqMP, causing issues as soon as the MPU is enabled for the respective target.
* Add the "zephyr,ocm" entry added to the devicetree to the list of supported "chosen" entries in the devicetree reference docs.

ibirnbaum committed:
dts: arm: xilinx: zynqmp: add OCM memory area

Add the 256 kB On-Chip memory area which is located at 0xfffc0000.
In conjunction with a corresponding 'chosen' entry at the board level,
equivalent to board device trees based on the Zynq-7000, explicit placement
of data in the OCM becomes possible, for example when setting up the GEM
Ethernet Controller's DMA area.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
boards: arm: qemu_cortex_r5: add chosen entry for OCM

Add the 'chosen' entry for the OCM whose declaration is being added to the
SoC's device tree. This enables the linker command file to properly place
data explicitly assigned ocm_bss/ocm_data sections.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: xlnx_gem: remove Zynq/ZynqMP distinction for DMA placement

With the unification of OCM declaration & assignment between Zynq-7000
and Ultrascale/ZynqMP, remove the distinction between those two SoC
families so that the DMA area is always set up in the OCM regardless
of the current SoC type.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

carlescufi commented:
@ibirnbaum please rebase

mbolivar-nordic commented on boards/arm/qemu_cortex_r5/qemu_cortex_r5.dts:
```diff
@@ -13,10 +13,11 @@
+		zephyr,flash      = &flash0;
+		zephyr,console    = &uart0;
 		zephyr,shell-uart = &uart0;
+		zephyr,ocm        = &ocm;
```
This must be documented in the `devicetree-chosen-nodes` section of the devicetree reference pages.

ibirnbaum replied:
I've added the documentation of the "zynq,ocm" chosen entry to the devicetree reference docs, comp. the newly added commit.

ibirnbaum committed:
doc: devicetree: add "zephyr,ocm" to list of "chosen" nodes

Add "zephyr,ocm" to the list of supported "chosen" nodes in
the devicetree reference docs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

stephanosio commented on boards/arm/qemu_cortex_r5/qemu_cortex_r5.dts:
```diff
@@ -13,10 +13,11 @@
-		zephyr,sram = &sram0;
-		zephyr,flash = &flash0;
-		zephyr,console = &uart0;
+		zephyr,sram       = &sram0;
+		zephyr,flash      = &flash0;
+		zephyr,console    = &uart0;
```
Not sure if this is the best thing to do; sure it looks nice, but such attempts can easily get out of hand and does not necessarily help with code readability.

ibirnbaum replied:
reverted the indentation changes.

stephanosio commented on doc/reference/devicetree/api.rst:
```diff
@@ -387,6 +387,8 @@ device.
        interprocess-communication (IPC)
    * - zephyr,itcm
      - Instruction Tightly Coupled Memory node on some Arm SoCs
+   * - zephyr,ocm
+     - On-chip memory node on Xilinx Zynq-7000 and ZynqMP SoCs
```
Should this be more generalised?

The "on-chip memory" in this context refers to the SRAM on the microprocessor chip itself as opposed to external DRAMs (e.g. DDR), and there are many SoCs that have such a construct -- of course, they could be named somewhat differently, but adding a new name for such a construct per SoC does not exactly sound like a sane thing to do.

ibirnbaum replied:
That's basically what I was expecting - that something as generic as an OCM, under this particular name, should be a common thing among various ARM SoCs. But it seems like the Xilinx SoCs are actually the first SoCs to have this feature. The existing types ITCM and DTCM make assumptions on the purpose of the integrated memory, which the OCM does not. ST at least uses a different name for their on-chip memory, I haven't looked into whether a specific purpose is implied for their CCM. So, at the time being, there don't seem to be any SoC families to generalise the OCM to. In case the addition of SoCs made by vendors other than those already supported is pending, I can check if those come with an identical OCM.
